ec/src/models/bls12/mod.rs

```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
             // r = f^((p^6 - 1)(p^2 + 1))
             r *= &f2;
 
-            // Hard part of the final exponentation is below:
-            // From https://eprint.iacr.org/2016/130.pdf, Table 1
+            // Hard part of the final exponentation:
+            // t[0].CyclotomicSquare(&result)
             let mut y0 = r.cyclotomic_square();
-            y0.conjugate();
-
-            let mut y5 = Self::exp_by_x(r);
-
-            let mut y1 = y5.cyclotomic_square();
-            let mut y3 = y0 * &y5;
-            y0 = Self::exp_by_x(y3);
-            let y2 = Self::exp_by_x(y0);
-            let mut y4 = Self::exp_by_x(y2);
-            y4 *= &y1;
-            y1 = Self::exp_by_x(y4);
-            y3.conjugate();
-            y1 *= &y3;
-            y1 *= &r;
-            y3 = r;
-            y3.conjugate();
-            y0 *= &r;
-            y0.frobenius_map(3);
-            y4 *= &y3;
-            y4.frobenius_map(1);
-            y5 *= &y2;
-            y5.frobenius_map(2);
-            y5 *= &y0;
-            y5 *= &y4;
-            y5 *= &y1;
-            y5
+            // t[1].Expt(&result)
+            let mut y1 = Self::exp_by_x(r);
+            // t[2].InverseUnitary(&result)
+            let mut y2 = r;
+            y2.conjugate();
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= y2;
+            // t[2].Expt(&t[1])
+            y2 = Self::exp_by_x(y1);
```

Comment:
I would be wary if copies of Fp12 elements, in my experience the compiler produce a bench of movaps, movups instructions, you might want an out-of-place pow_by_x

Reply:
To clarify, you mean that `exp_by_x` should take `x` by reference? Or?

Reply:
I mean mutate `y2` in-place